example/client/burl/cookie.cpp

```diff
@@ -0,0 +1,346 @@
+            c != ',' && c != ';'  && c != ':' && c != '\\' && c != '"' &&
+            c != '/' && c != '['  && c != ']' && c != '?'  && c != '=' &&
+            c != '{' && c != '}';
+    }
+};
+
+constexpr auto name_chars = name_chars_t{};
+
+struct value_chars_t
+{
+    constexpr
+    bool
+    operator()(char c) const noexcept
+    {
+        return
+            (c == 0x21             ) ||
+            (c >= 0x23 && c <= 0x2B) ||
+            (c >= 0x2D && c <= 0x3A) ||
+            (c >= 0x3C && c <= 0x5B) ||
+            (c >= 0x5D && c <= 0x7E);
+    }
+};
+
+constexpr auto value_chars = value_chars_t{};
+
+constexpr auto attr_chars =
+    urls::grammar::all_chars -
+    urls::grammar::lut_chars("\x1F\x7f;");
+
+bool
+domain_match(
+    core::string_view r_domain,
+    core::string_view c_domain,
+    bool subdomains) noexcept
+{
+    if(!subdomains)
+        return r_domain == c_domain;
+
+    if(c_domain.starts_with('.'))
+        c_domain.remove_prefix(1);
+
+    if(r_domain.ends_with(c_domain))
+    {
+        if(r_domain.size() == c_domain.size())
+            return true;
+
+        return r_domain[r_domain.size() - c_domain.size() - 1] == '.';
+    }
+
+    return false;
+}
+
+bool
+path_match(
+    core::string_view r_path,
+    core::string_view c_path) noexcept
+{
+    if(r_path.empty())
+        return true;
+
+    if(r_path.starts_with(c_path))
+    {
+        if(r_path.size() == c_path.size())
+            return true;
+
+        if(c_path.ends_with('/'))
+            return true;
+
+        return r_path[r_path.size() - c_path.size()] == '/';
+    }
+
+    return false;
+}
+
+} // namespace
+
+boost::system::result<cookie>
+parse_cookie(core::string_view sv)
+{
+    namespace ug = urls::grammar;
+
+    const auto parse_rs = ug::parse(
+        sv,
+        ug::tuple_rule(
+            ug::token_rule(name_chars),
+            ug::squelch(ug::delim_rule('=')),
+            ug::optional_rule(ug::token_rule(value_chars)),
+            ug::range_rule(
+                ug::tuple_rule(
+                    ug::squelch(ug::delim_rule(';')),
+                    ug::squelch(ug::optional_rule(ug::delim_rule(' '))),
+                    ug::token_rule(attr_chars - ug::lut_chars('=')),
+                    ug::squelch(ug::optional_rule(ug::delim_rule('='))),
+                    ug::optional_rule(ug::token_rule(attr_chars))))));
+
+    if(parse_rs.has_error())
+        return parse_rs.error();
+
+    auto rs  = cookie{};
+    rs.name  = std::get<0>(parse_rs.value());
+    rs.value = std::get<1>(parse_rs.value());
+
+    for( auto&& attr : std::get<2>(parse_rs.value()))
+    {
+        auto name  = std::get<0>(attr);
+        auto value = std::get<1>(attr);
+
+        if(ug::ci_is_equal(name, "Expires"))
+        {
+            if(!value)
+                return ug::error::invalid;
+            // TODO: There are more date formats; we need a
+            // better parsing method.
```

Comment:
I think this code should be factored out into its own function with a good signature. You can still call into `std` for the implementation, which is a hack yet it is an acceptable hack for now since it lets us move forward without writing a big piece of date parsing code. And having it in a separate function will keep this function clean.